formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonTreeReader.kt

```diff
@@ -75,9 +83,33 @@ internal class JsonTreeReader(
         return when (lexer.peekNextToken()) {
             TC_STRING -> readValue(isString = true)
             TC_OTHER -> readValue(isString = false)
-            TC_BEGIN_OBJ -> readObject()
+            TC_BEGIN_OBJ -> {
+                /*
+                 * If the object has the depth of 200 (an arbitrary "good enough" constant), it means
+                 * that it's time to switch to stackless recursion to avoid StackOverflowError.
+                 * This case is quite rare and specific, so more complex nestings (e.g. through
+                 * the chain of JsonArray and JsonElement) are not supported.
+                 */
+                val result = if (++stackDepth == 200) {
+                    readDeepRecursive()
+                } else {
+                    readObject()
```

Comment:
I may be missing something here, but how do we actually make a recursive call when we're on e.g. stack depth 202 (object inside an object that triggered callRecursive)

Reply:
On the depth of 200 we switch to `readDeepRecursive`.
`readDeepRecursive` is a recursive function that invokes `DeepRecursiveScope<Unit, JsonElement>.readObject()` which, when parsing, invokes `callRecursive` (aka `readDeepRecursive`). The stack depth is no longer maintained as soon as we're in `readDeepRecursive`